src/permutation.nr

```diff
@@ -2,527 +2,234 @@ use crate::encoding::{
     decode_choose, decode_majority, decode_xor, encode_a, encode_e, encode_message_extension,
     EncodedChoose, EncodedMajority, EncodedWitness,
 };
-
-// Size of the message block when packed as 8-byte integer array.
-global INT_BLOCK_SIZE: u32 = 16;
-
-type INT_BLOCK = [u64; INT_BLOCK_SIZE];
-
-// The state accumulates the blocks.
-// Its overall size is the same as the `HASH`.
-type STATE = [u64; 8];
-
-fn extend_witness(witness: [Field; 16]) -> [Field; 80] {
+use crate::formatting::reduce_modulo_64;
+use crate::tables::SHA512_ROUND_CONSTANTS;
+
+/// Represent the SHA384/512 permutation state as base4-encoded variables
+pub(crate) struct RoundState {
+    pub(crate) a: EncodedMajority,
+    pub(crate) b: EncodedMajority,
+    pub(crate) c: EncodedMajority,
+    pub(crate) d: EncodedMajority,
+    pub(crate) e: EncodedChoose,
+    pub(crate) f: EncodedChoose,
+    pub(crate) g: EncodedChoose,
+    pub(crate) h: EncodedChoose,
+}
+
+/// Perform message extension, expanding 16 words of the message schedule into 80 words
+pub(crate) fn extend_message(witness: [Field; 16]) -> [Field; 80] {
     let mut w: [EncodedWitness; 80] = [EncodedWitness::new(); 80];
+
     for i in 0..16 {
         w[i] = encode_message_extension(witness[i]);
     }
-    // note for improvements:
-    // 1: we don't need to re-encode for `i = 79, i = 80`
-    // 2: we only need partial re-encoding for i = [65...,78]
-    // 3: we only need partial encoding for i = [0,..,13]
-    for i in 16..80 {
-        let s0 = decode_xor(w[i - 15].ror1 + w[i - 15].ror8 + w[i - 15].rshift7);
+    for i in 16..78 {
+        // s0 = (w[i-15] rightrotate 7) xor (w[i-15] rightrotate 18) xor (w[i-15] rightshift  3)
+        // When we encoded `w`, w15_rotate_sum represents the base4-encoded linear sum of the above
+        // all we need to do is decode back into base2
+        let s0: Field = decode_xor(w[i - 15].w15_rotate_sum);
 
-        let s1 = decode_xor(w[i - 2].ror19 + w[i - 2].ror61 + w[i - 2].rshift6);
+        // s1 = (w[i-2] rightrotate 17) xor (w[i-2] rightrotate 19) xor (w[i-2] rightshift 10)
+        // When we encoded `w`, w2_rotate_sum represents the base4-encoded linear sum of the above
+        // all we need to do is decode back into base2
+        let s1 = decode_xor(w[i - 2].w2_rotate_sum);
 
+        // evaluating the sum as Field elements and then reducing modulo 64 is more efficient than `u64` arithmetic
         let new_witness = reduce_modulo_64(w[i - 16].raw + s0 + w[i - 7].raw + s1);
 
+        // take the output and encode into the extended base4 representation
         w[i] = encode_message_extension(new_witness);
     }
 
     let mut result: [Field; 80] = [0; 80];
-    for i in 0..80 {
+    for i in 0..78 {
```

Comment:
small optimisation, we can skip `encode_message_extension` for final 2 iterations